src/ecfeed.hpp

```diff
@@ -0,0 +1,1064 @@
+            }
+        };
+
+        bool _done;
+        std::vector<T> _data;
+        const_iterator _begin;
+        const_iterator _end;
+        std::mutex _mutex;
+        std::mutex _cv_mutex;
+        std::condition_variable _cv;
+
+        MethodInfo _method_info;
+        bool _method_info_ready;
+
+    public:
+
+        typedef T value_type;
+
+        TestQueue() :
+            _done(false),
+            _begin(*this),
+            _end(*this, true),
+            _method_info_ready(false)
+        {}
+
+        const_iterator begin() const {return _begin;}
+        const_iterator end() const {return _end;}
+
+        friend class const_iterator;
+        friend class TestProvider;
+
+        bool done() {
+            std::unique_lock<std::mutex> cv_lock(_mutex);
+
+            if (_data.size() != 0) {
+                return false;
+            }
+            else if (_done) {
+                return true;
+            }
+
+            _cv.wait(cv_lock);
+            return _done;
+        }
+
+        void next() 
+        {
+            std::lock_guard<std::mutex> lock(_mutex);
+            _data.erase(_data.begin());
+        }
+
+        void insert(const T& element)
+        {
+            std::lock_guard<std::mutex> lock(_mutex);
+
+            _data.push_back(element);
+            _cv.notify_one();
+        }
+
+        T& current_element() 
+        {
+            std::unique_lock<std::mutex> cv_lock(_mutex);
+            if (_data.size() > 0) {
+                return _data[0];
+            }
+            _cv.wait(cv_lock);
+            return _data[0];
+        }
+
+        void finish()
+        {
+            std::lock_guard<std::mutex> lock(_mutex);
+            _done = true;
+            _cv.notify_one();
+        }
+
+        bool empty()
+        {
+            std::lock_guard<std::mutex> lock(_mutex);
+            return _data.size() == 0;
+        }
+
+        std::vector<T> toList() 
+        {
+            std::unique_lock<std::mutex> cv_lock(_mutex);
+
+            while (!_done) {
+                _cv.wait(cv_lock);
+            }
+
+            return _data;
+        }
+
+        std::vector<std::string> getArgumentTypes()
+        {
+            std::unique_lock<std::mutex> cv_lock(_mutex);
+
+            while (!getMethodInfoReady()) {
+                _cv.wait(cv_lock);
+            }
+
+            return _method_info.arg_types;
+        }
+        
+        std::vector<std::string> getArgumentNames()
+        {
+            std::unique_lock<std::mutex> cv_lock(_mutex);
+
+            while (!getMethodInfoReady()) {
+                _cv.wait(cv_lock);
+            }
+
+            return _method_info.arg_names;
+        }
+
+    private:
+
+        void setMethodInfoReady()
+        {
+            _method_info_ready = true;
+        }
+
+        bool& getMethodInfoReady() 
+        {
+            return _method_info_ready;
+        }
+
+        MethodInfo& getMethodInfo()
```

Comment:
API for other runners includes 'getArgumentNames' and 'getArgumentTypes'. It's not needed to reveal the whole structure (it is used internally). However, if you want it can be made public.

Reply:
you are correct - keep it like this